### DISCUSSION TITLE
Multilingual URLs

### DISCUSSION THREAD

**Discussion by motiejunas** [vercel/next.js#12412] [Help]
First time trying to play with next.js and I can't seem to find a proper solution to have multilingual website/routes.

Let's imagine I want to support English and German languages. Just a note: the set of languages should be asynchronously fetched from the third-party service, e.g. Contentful.

This is the URL structure I want to have:

> /
> /en
> /de
> /posts/1
> /en/posts/1
> /de/posts/1

If I do use the following folder structure:

> /pages/[lang]/index.tsx
> /pages/[lang]/posts/[id].tsx

Then I can't access master language routes: "/" and "/posts/1".

Alternatively I tried to define these routes in "next.config.js", but in that case:
1) can't use TypeScript
2) did not find a way to identify the language in "getStaticProps"

Any tips, examples?

**Comment by evantahler** [comment]
This thread may have what you are looking for https://github.com/zeit/next.js/issues/9081.  There is an experimental next.js feature `rewrites` where you could rewrite the un-localed' pages to the default locale

**Reply by motiejunas** [reply]
Does not seem to work either for static websites:
https://github.com/zeit/next.js/issues/9081#issuecomment-620487708

**Comment by landsman** [comment]
@motiejunas Hey, check our brand new solution for this:
https://github.com/trisbee/next-i18n-routes-middleware

Would you please give us feedback?

**Comment by Vadorequest** [comment]
@motiejunas @landsman Check out this demo:
https://nrn-v1-hyb-mst-aptd-gcms-lcz-sty-c1-akenkjv4z.now.sh/en-US/terms

4 different locales are available (fr, fr-FR, en, en-US) for testing. Only 2 are available through the footer button.

Source code is here:
https://github.com/UnlyEd/next-right-now/pull/42#issuecomment-623145867

It has reached a good maturity level, I'm confident in the implementation.

It relies on experimental features rewrites/redirects.

**Reply by landsman** [reply]
@Vadorequest Where can I find your routes specification? 

I think that using exported object with routes is better for testing, prevent 404 etc, less duplicate strings = more stable product.

We are using these paths in navigation: https://github.com/trisbee/next-i18n-routes-middleware/blob/master/examples/simple/server/routes.js#L6-L11 which turns as pretty good. Of course with custom component for link like: `<Link href={paths.homepage}>menu item translated with lingui</Link>`.

Did you think about something like that?

**Reply by Vadorequest** [reply]
Everything lives in this branch:
https://github.com/UnlyEd/next-right-now/blob/v1-hyb-mst-aptd-gcms-lcz-sty/next.config.js#L33-L75

**Comment by leerob** [comment]
Check out the new [i18n RFC](https://github.com/vercel/next.js/discussions/17078) and give us your feedback! 🚀

**Reply by Vadorequest** [reply]
@leerob Note that multilingual URLS aren't part of RFC as of now, though.

**Comment by anialamo** [comment]
Any solution for this case Multilingual URLs?
Thx!
